### ISSUE TITLE
Format with empty/blank values does not coerce to nil

### ISSUE THREAD

**Issue by hlegius** [hanami/validations#90]
Hello,

As described by @wafcio in [his gist](https://gist.github.com/wafcio/164b1191d2d689bfa088), he was trying to validate a value with `format` attribute. If pass an empty string, format returns `valid true` (which is correct), but does not coerce into `nil`. If we provide `type: Integer` attribute, it will return 0, which isn't the same of nil of blank.

```ruby
class Person
  include Hanami::Validations

  attribute :number, format: /\A\d+\z/
end

person = Person.new(number: '')
p.valid? # => true
```

I was wondering if we should introduce blank/empty string to nil conversion in all occasions as well.

**Comment by jodosha** [comment]
@hlegius Hello and thanks for taking care of this. Your suggestion is good and can work. If it won't take you too much time please implement it. :+1: 

In the long run, we should probably add two different layers of validations. A "HTTP" layer that solves the problem of "everything is a string", which can be used for `Action::Params` and this kind of problems reported by @wafcio.

A second layer for "core" types, where we're not trying to guess the input and coerce it, but only to verify that the given data is already of the expected type.

---

How this can solve the problem described? As now, the coercion of the types is implemented by `Hanami::Validations::Coercions`, and under the hood by `Hanami::Utils::Kernel`. The semantic of the latter is: "try to do your best to convert this input into the target type, otherwise raise a `TypeError`". But for `Validations` we suppress that error and return `nil`.

This implementation is "too smart": it tries to do the conversion, but then gives up and returns `nil`. So for an input `""` (empty string), we internally read a `nil` value. We skip validations on `nil` values and here's the source of our problem.

Now, this mechanism is in place because it tries to solve two kind of problems: HTTP and Core types in at the same time.

---

**My suggestion is to decouple them**. When an object includes `Hanami::Validations` it only gets the "core" set of coercion rules. Do we expect an `Integer`? Can be the input safely converted? If so, do it, otherwise complain. This is the logic of `Utils::Kernel`.

We should then introduce another module like `Hanami::Validations::Http`, which is included by default by `Action::Params` and it tries to handle string inputs and than make use of the core layer.

---

This can fix the root problem, when validations are used on objects, without the implications of coercible HTTP params.

**Comment by hlegius** [comment]
@jodosha 

> Your suggestion is good and can work. If it won't take you too much time please implement it. :+1: 

I was wondering if #40 isn't the opposite of the problem I've described above. Maybe if we coerce empty string to `nil` will break again that issue #40. Or, maybe I've just don't get what went wrong on that issue. Suggestions?

In both cases (I am just crazy and that issue is not the opposite) or we need to think a bit more, I can handle it. o/

> My suggestion is to decouple them. When an object includes Hanami::Validations it only gets the "core" set of coercion rules. Do we expect an Integer? Can be the input safely converted? If so, do it, otherwise complain. This is the logic of Utils::Kernel.

> We should then introduce another module like Hanami::Validations::Http, which is included by default by Action::Params and it tries to handle string inputs and than make use of the core layer.

:+1: :100: 
It will be a interesting improvement. Personally, will be very handy. While using H::V, sometimes I miss exactly what you described: input validation (tainted data from user input)/type-coerce validation. I'd vote for a proposal on this one.

**Comment by solnic** [comment]
In dry-validation this problem has been solved by giving you an explicit way of specifying that a value can be nil, ie:

``` ruby
schema = Dry::Validation.Form do
  key(:number).maybe(:number?)
end

schema.(number: "").to_h # { number: nil }
schema.(number: "1").to_h # { number: 1 }
schema.(number: "ooops").messages # { number: ["must be a number"] }
```

Notice that it properly coerces the value based on your expectation.

**Comment by jodosha** [comment]
Closed by #100
